codec/src/main/java/io/netty/handler/codec/ByteString.java

```diff
@@ -0,0 +1,770 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.handler.codec;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufProcessor;
+import io.netty.buffer.Unpooled;
+import io.netty.util.CharsetUtil;
+import io.netty.util.internal.StringUtil;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Comparator;
+
+/**
+ * The primary use case for this class is to function as an immutable array of bytes. For performance reasons this
+ * class supports sharing memory with external arrays, and also direct access to the underlying memory via
+ * {@link #array()}. Care must be taken when directly accessing the memory as that may invalidate assumptions that
+ * this object is immutable.
+ */
+public class ByteString {
```

Comment:
General comment, since you allow `ByteString` to be extended, would it make sense to mark many of the frequently called methods as `final` to help with optimization?